### ISSUE TITLE
Bug: [@typescript-eslint/no-unused-vars] Picking up `key` in `key in Union` as an unused variable

### ISSUE THREAD

**Issue by jonahallibone** [typescript-eslint/typescript-eslint#10028]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

I was unable to find matching versions 

I am on 
```
"@typescript-eslint/eslint-plugin": "^8.6.0",
"@typescript-eslint/parser": "^8.6.0",
"typescript": "^5.6.2"
```

### Repro Code

```TypeScript
/**
 * Interface defining the general table variant
 */
export type GeneralTableVariant = "clean" | "grid";

/**
 * Types for general table style context
 */
interface TableStyleContextProps {
  variant: GeneralTableVariant;
  rowHeightPx: `${number}px`;
  rowHeight: number;
}

export type GeneralTableVariantStyles = {
  [key in GeneralTableVariant]: SystemProps;
};
```

### ESLint Config

```javascript
eslintrc.js


const sharedRules = {
  curly: ["error", "all"],
  "react/jsx-filename-extension": [
    "warn",
    {
      extensions: [".js", ".tsx"],
    },
  ],
  "@tanstack/query/exhaustive-deps": "error",
  "@tanstack/query/no-rest-destructuring": "warn",
  "@tanstack/query/stable-query-client": "error",
  "react/jsx-props-no-spreading": "off",
  "react/function-component-definition": [
    "error",
    {
      namedComponents: "arrow-function",
      unnamedComponents: "arrow-function",
    },
  ],
  "react/require-default-props": "off",
  "react/react-in-jsx-scope": "off",
  "import/prefer-default-export": "off",
  "import/no-cycle": [
    "warn",
    {
      disableScc: true,
      ignoreExternal: true,
    },
  ],
  "import/no-extraneous-dependencies": [
    "error",
    {
      devDependencies: [
        "**/*.spec.ts",
        "**/*.spec.tsx",
        "**/playwright.config.ts",
      ],
    },
  ],
  "import/extensions": [
    "error",
    {
      js: "never",
      ts: "never",
      tsx: "never",
    },
  ],
  "no-param-reassign": ["error", { props: false }],
};

module.exports = {
  env: {
    browser: true,
    es6: true,
    mocha: true,
    jest: true,
  },
  extends: [
    "eslint:recommended",
    "plugin:react/recommended",
    "next/core-web-vitals",
    "airbnb",
    "airbnb/hooks",
    "prettier",
    "plugin:@tanstack/eslint-plugin-query/recommended",
  ],
  globals: {
    Atomics: "readonly",
    SharedArrayBuffer: "readonly",
    window: true,
    document: true,
    localStorage: true,
    FormData: true,
    FileReader: true,
    Blob: true,
    navigator: true,
    cy: true,
    Cypress: true,
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
    ecmaVersion: "latest",
    sourceType: "module",
  },
  root: true,
  rules: sharedRules,
  overrides: [
    // TypeScript files
    {
      files: ["**/*.ts", "**/*.tsx"],
      parser: "@typescript-eslint/parser",
      parserOptions: {
        ecmaVersion: "latest",
        project: ["./tsconfig.json"],
        projectService: true,
        tsconfigRootDir: __dirname,
      },
      plugins: [
        "deprecation",
        "tsdoc",
        "@typescript-eslint",
        "@tanstack/query",
      ],
      extends: [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:react/recommended",
        "airbnb",
        "airbnb-typescript",
        "next/core-web-vitals",
        "airbnb/hooks",
        "prettier",
      ],
      rules: {
        ...sharedRules,
        "deprecation/deprecation":
          process.env.NODE_ENV === "production" ? "off" : "warn",
        "tsdoc/syntax": "warn",
        "react/jsx-key": "error",
        "react/display-name": "off",
        "@typescript-eslint/ban-ts-comment": "warn",
        "@typescript-eslint/no-explicit-any": "warn",
        "no-unused-vars": "off",
        "@typescript-eslint/no-unused-vars": "error",
        "@typescript-eslint/consistent-type-imports": [
          "warn",
          {
            prefer: "type-imports",
            disallowTypeAnnotations: true,
          },
        ],
        "import/extensions": 0, // maybe this can be better configured
      },
    },
    // JavaScript files
    {
      files: ["**/*.js"],
      plugins: ["jsdoc"],
      extends: [
        "eslint:recommended",
        "airbnb",
        "airbnb/hooks",
        "next/core-web-vitals",
        "plugin:jsdoc/recommended",
        "prettier",
      ],
      rules: {
        ...sharedRules,
        "jsdoc/require-jsdoc": "off",
      },
    },
    // JavaScript config files & Script files
    {
      files: ["./*.js", "./scripts/*.js"],
      rules: {
        "no-console": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
    // Cypress tests
    {
      files: ["./cypress/**/*.ts"],
      parserOptions: {
        project: "./cypress/tsconfig.json",
        tsconfigRootDir: __dirname,
      },
    },
    {
      files: ["./tests/**/*.ts"],
    },
  ],
};
```

### tsconfig

```json
{
  "compilerOptions": {
    "baseUrl": "src",
    "target": "esnext",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "noImplicitAny": true,
    "skipLibCheck": true,
    "strict": true,
    "strictNullChecks": true,
    "forceConsistentCasingInFileNames": true,
    "noUncheckedIndexedAccess": true,
    "noEmit": true,
    "incremental": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "typeRoots": ["../node_modules/@types", "./types"],
    "plugins": [
      {
        "name": "next"
      }
    ]
  },
  "ts-node": {
    "transpileOnly": true,
    "compilerOptions": {
      "module": "CommonJS",
      "moduleResolution": "node"
    }
  },
  "include": [
    "next-env.d.ts",
    "*.js",
    "**/*.ts",
    "**/*.tsx",
    "types",
    "index.d.ts",
    "src",
    "tests",
    ".next/types/**/*.ts"
  ],
  "exclude": ["node_modules", ".next"]
}
```

### Expected Result

`key` should not be picked up by the linter

### Actual Result

Key is picked up as `unused`
<img width="694" alt="Screenshot 2024-09-20 at 11 53 31 AM" src="https://github.com/user-attachments/assets/aae4d54f-9e05-46ce-8bc7-803c81f8cd52" alt="error image">
